lib/config/applyrule.cpp

```diff
@@ -131,23 +133,33 @@ bool ApplyRule::HasMatches() const
 	return m_HasMatches;
 }
 
-std::vector<ApplyRule>& ApplyRule::GetRules(const String& type)
+std::vector<ApplyRule>& ApplyRule::GetRules(const Type::Ptr& sourceType, const Type::Ptr& targetType)
 {
-	auto it = m_Rules.find(type);
-	if (it == m_Rules.end()) {
-		static std::vector<ApplyRule> emptyList;
-		return emptyList;
+	auto perSourceType (m_Rules.find(sourceType.get()));
+
+	if (perSourceType != m_Rules.end()) {
+		auto perTargetType (perSourceType->second.find(targetType.get()));
+
+		if (perTargetType != perSourceType->second.end()) {
+			return perTargetType->second;
+		}
 	}
-	return it->second;
+
+	static std::vector<ApplyRule> noRules;
+	return noRules;
```

Comment:
For the sake of more readable code, I'd say `return m_Rules[sourceType.get()][targetType.get()];` would also be perfectly fine here. Worst case it would allocate like 10 extra map entries.

Reply:
No, that would be the best case. Worst case it would allocate the extra map entries concurrently. \*CRASH\*

Reply:
Ah, the fun of what runs concurrently in Icinga and what doesn't :upside_down_face: